File: ruby/sieve/sieve.rb
class Sieve
  def initialize(n)
    @max = n
    @marks = Array.new(@max, false)
  end

  def primes
    2.upto(@max).each_with_object([]) do |i, primes|
      next if @marks[i]

      primes << i
      mark(i)
    end
  end

  private

  def mark(x)
    x.step(@max, x) { |i| @marks[i] = true }
  end
end

File: scala/sieve/src/main/scala/Sieve.scala
object Sieve {

  def addMarks(marks: Set[Int], x: Int, max: Int): Set[Int] = {
    marks ++ x.to(max, x)
  }

  def primesUpTo(n: Int): Seq[Int] = {
    if (n == 1) return List()

    var marks = Set[Int]()
    var primes = List[Int]()
    2.to(n).foreach(x => {
      if (!marks.contains(x)) {
        marks = addMarks(marks, x, n)
        primes = primes :+ x
      }
    })
    primes
  }

}
